### PR TITLE
Chplquerier2

### PR DESCRIPTION
**Pull request by edeyoung**
This PR adds the ability to query CHPL for its HealthIT products.
Several other changes were made in support of this PR including:
* Adding an "Update" method for HealthITProduct that allows all the fields of a HealthITProduct to be updated with the fields of another HealthITProduct except the database ID, creation time, and update time. This is to support updating the actual health it product in the database (so it retains its db id).
* Adding `context` to the `endpointmanager *Store` interface calls (e.g., HealthITProductStore.AddHealthITProduct), updating all structs implementing those calls, and updating the postgres db access calls to use `context`.
* Adding a basic mock implementation for HealthITProductStore. Note: only the mocks for the functions I wanted to use were created. If decided it's necessary, I can update the mocks for everything. I had some frustrations with how I organized everything to be a different type of *Store, but that seemed like to much of an overhaul to change here.
* Adding testhelpers, which includes an "Assert" function and a way to mock an HTTP server through the HTTP client.
* log to terminal instead of to file, which is best practice for dockerized services.

To run this so it goes out and gets products from CHPL and stores them:
* make sure your DB is running, all your environment variables are set for accessing the DB, and the tables have been created within the DB. See the README file for information on how to do this.
* make sure you have the CHPL API key set in your environment variables. This should be the API key noted here: https://mitre.sharepoint.com/sites/ProjectLantern/SitePages/Project-Resources.aspx?source=https://mitre.sharepoint.com/sites/ProjectLantern/SitePages/Forms/ByAuthor.aspx
* run the code following the instructions in the endpointmanager/README.md

----

Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:**
- [x] Administrative tasks are complete.
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is: https://oncprojectracking.healthit.gov/support/browse/LANTERN-76
  - vendor/ directories have been updated
    - New dependencies have been added
    - Local dependencies have been updated
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
- [x] Code review has been performed.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

- [x] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [x] Code review has been performed.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
  - vendor/ directories have been updated
    - New dependencies have been added
    - Local dependencies have been updated
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Secondary Reviewer:**

- [x] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [x] Code review has been performed.
  - Code accomplishes the tasks purpose.
- [x] Tests are complete.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge is complete.

### CODE REVIEW

File: endpointmanager/pkg/chplquerier/productquerier.go [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+// takes the json byte string and converts it into the associated JSON model
+func convertProductJSONToObj(ctx context.Context, prodJSON []byte) (*chplCertifiedProductList, error) {
+	var prodList chplCertifiedProductList
+
+	// don't unmarshal the JSON if the context has ended
+	select {
+	case <-ctx.Done():
+		return nil, errors.Wrap(ctx.Err(), "Unable to convert product JSON to objects - context ended")
+	default:
+		// ok
```
Comment (bmath10) [inline endpointmanager/pkg/chplquerier/productquerier.go]: Is there a reason why the default part of the select statement is kept here and elsewhere (like line 199)? It seems like you could just get rid of it, but I could see it being a placeholder for some handling you were considering for the future!

Comment (edeyoung) [reply endpointmanager/pkg/chplquerier/productquerier.go]: the `default` is necessary. without it, the `select` statement will wait until a message shows up in one of its case statements.

File: endpointmanager/pkg/chplquerier/productquerier.go [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+	// cert dates are the same. checking certification criteria lists. if new prod has more criteria, should update.
+	if len(existingDbProd.CertificationCriteria) < len(newDbProd.CertificationCriteria) {
+		return true, nil
+	} else if len(existingDbProd.CertificationCriteria) > len(newDbProd.CertificationCriteria) {
+		return false, nil
+	}
+
+	// certification criteria lists are the same lengths. unknown precedent for updates. return error.
+	if !cmp.Equal(existingDbProd.CertificationCriteria, newDbProd.CertificationCriteria) {
+		return false, errors.New("HealthITProducts certification edition and date are equal, but has same number but unequal certification criteria - unknown precendence for updates")
```
Comment (bmath10) [inline endpointmanager/pkg/chplquerier/productquerier.go]: This error message is kind of confusing to me. is it saying that the number of certification criteria are equal but they are different criteria?

File: endpointmanager/pkg/chplquerier/productquerier.go (hunk 1 above)
Comment (bmath10) [reply endpointmanager/pkg/chplquerier/productquerier.go]: ah okay, i guess i misunderstood how the select statement worked, interesting!

**Comment by bmath10** [comment]
Just out of curiosity, why was the logging config removed?

**Comment by bmath10** [comment]
My one issue while running: in the endpointmanager readme, the final line says ```go run main.go``` and I had to run ```go run cmd/main.go```

**Comment by edeyoung** [comment]
Apparently it's best practice to log to the terminal for microservices. Also, I was getting frustrated not seeing the log messages in the terminal while I was debugging stuff. So best practice + easier for me. Thank you for noting that! I'll add it to my PR notes.

File: endpointmanager/cmd/main.go [hunk 3]
```diff
@@ -43,28 +45,24 @@ func setupConfig() {
...
 
 	store, err := postgresql.NewStore(viper.GetString("dbhost"), viper.GetInt("dbport"), viper.GetString("dbuser"), viper.GetString("dbpass"), viper.GetString("dbname"), viper.GetString("dbsslmode"))
 	if err != nil {
 		panic(err.Error())
 	}
 	defer store.Close()
 	fmt.Println("Successfully connected!")
+
+	ctx := context.Background()
+	client := &http.Client{}
```
Comment (mayerm94) [inline endpointmanager/cmd/main.go]: should add a timeout to the default client https://medium.com/@nate510/don-t-use-go-s-default-http-client-4804cb19f779

**Pull request merged**